solr/contrib/ltr/src/java/org/apache/solr/ltr/LTRThreadModule.java

```diff
@@ -15,69 +15,139 @@
+      args.remove(CONFIG_PREFIX+key);
+    }
+
+    return extractedArgs;
   }
-   
-   public int getMaxThreads(){
-      return maxThreads;
-   }
-   
-   public int getMaxQueryThreads(){
-     return maxQueryThreads;
-   }
-   
-   public Semaphore getLTRSemaphore(){
-     return ltrSemaphore;
-   }
-   
-   public Executor getCreateWeightScoreExecutor(){
-     return createWeightScoreExecutor;
-   }
+
+  // settings
+  private int maxThreads = 1;
+  private int maxQueryThreads = 1;
+  private int maxPoolSize = Integer.MAX_VALUE;
+  private long keepAliveTimeSeconds = 10;
+  private String threadNamePrefix = "ltrExecutor";
+
+  ThreadMXBean tmxb = ManagementFactory.getThreadMXBean();
+
+  // implementation
+  private Semaphore ltrSemaphore;
+  private Executor createWeightScoreExecutor;
+
+  private final void initCreateWeightScoreExecutor() {
+  }
+
+  public LTRThreadModule() {
+  }
+
+  // For test use only.
+  LTRThreadModule(int maxThreads, int maxQueryThreads) {
+    this.maxThreads = maxThreads;
+    this.maxQueryThreads = maxQueryThreads;
+    init(null);
+  }
+
+  @Override
+  public void init(NamedList args) {
+    if (args != null) {
+      SolrPluginUtils.invokeSetters(this, args);
+    }
+    validate();
+    if  (this.maxThreads > 1 ){
+      ltrSemaphore = new Semaphore(maxThreads);
+    } else {
+      ltrSemaphore = null;
+    }
+    createWeightScoreExecutor = new ExecutorUtil.MDCAwareThreadPoolExecutor(
+        0,
+        maxPoolSize,
+        keepAliveTimeSeconds, TimeUnit.SECONDS, // terminate idle threads after 10 sec
+        new SynchronousQueue<Runnable>(),  // directly hand off tasks
+        new DefaultSolrThreadFactory(threadNamePrefix)
+        );
+  }
+
+  public void validate() {
+    if (maxThreads <= 0){
+      throw new IllegalArgumentException("maxThreads cannot be less than 1");
+    }
+    if (maxQueryThreads <= 0){
+      throw new IllegalArgumentException("maxQueryThreads cannot be less than 1");
+    }
+    if (maxThreads < maxQueryThreads){
+      throw new IllegalArgumentException("maxQueryThreads cannot be greater than maxThreads");
+    }
+  }
+
+  public void setMaxThreads(int maxThreads) {
+    this.maxThreads = maxThreads;
+  }
+
+  public void setMaxQueryThreads(int maxQueryThreads) {
+    this.maxQueryThreads = maxQueryThreads;
+  }
+
+  public void setMaxPoolSize(int maxPoolSize) {
+    this.maxPoolSize = maxPoolSize;
+  }
+
+  public void setThreadNamePrefix(String threadNamePrefix) {
+    this.threadNamePrefix = threadNamePrefix;
+  }
+
+  public Semaphore createQuerySemaphore() {
+    return (maxQueryThreads > 1 ? new Semaphore(maxQueryThreads) : null);
+  }
+
+  public void acquireLTRSemaphore() throws InterruptedException {
+    ltrSemaphore.acquire();
+  }
+
+  public void releaseLTRSemaphore() throws InterruptedException {
+    ltrSemaphore.release();
+  }
+
+  public void execute(RunnableFuture<FeatureWeight> runnableFuture) {
```

Comment:
No need for dependency on FeatureWeght, can just take in a Runnable instead

void execute(Runnable command);